### PR TITLE
Use glimmer tracked

### PR DESCRIPTION
**Pull request by tchak**

### CODE REVIEW

**Comment by tchak** [comment]
I am a bit stuck here. This works in octan ember (decorators) and classic ember (computed). But I am a bit lost at how to make decorators backward compatible as in been able tu use them as computed in new ember... Is there a way without explicitly wrapping them in `computed`? Do I have a way to determine if a macro is called as a computed or as a decorator? cc @pzuraq

**Comment by tchak** [comment]
So the current status of this is that it works in all versions of ember supported by `ember-orbit`. The only thing not working is usage of computed properties in ember > 3.15. If there is no simple runtime detection technique I think adding a config flag `EMBER_ORBIT_USE_TRACKED` is reasonable to help people transition.

**Comment by dgeb** [comment]
I'm fine with only supporting `tracked` in v0.17 and later of ember-orbit. I'd prefer not include the weight of the code in `classic-model` (and its associated maintenance cost), even if it means that folks need to upgrade to octane in order to benefit from the latest features in ember-orbit.

Do you have strong feelings about this, @tchak?

**Comment by tchak** [comment]
I agree @dgeb. Will rebase without legacy model. It will be much simpler and readable.

File: addon/-private/fields/attr.ts [hunk 1]
```diff
@@ -1,24 +1,62 @@
...
+    let { get: originalGet, set: originalSet } = trackedDesc;
+
+    let defaultAssigned = new WeakSet();
+
+    function setDefaultValue(record: Model) {
+      let value = record.getAttribute(key);
+      setValue(record, value);
+    }
+
+    function setValue(record: any, value: any) {
```
Comment (dgeb) [inline addon/-private/fields/attr.ts:18]:
```suggestion
    function setValue(record: Model, value: any) {
```

Won't `record` always be typed as a `Model`?

File: addon/-private/fields/attr.ts [hunk 2]
```diff
@@ -1,24 +1,62 @@
...
+      defaultAssigned.add(record);
+      return originalSet!.call(record, value);
+    }
+
+    function get(this: Model) {
```
Comment (dgeb) [inline addon/-private/fields/attr.ts:23]:
```suggestion
    function get(record: Model) {
```

Wouldn't it be better to keep the argument names consistent?

File: addon/-private/fields/attr.ts [hunk 3]
```diff
@@ -1,24 +1,62 @@
...
+      }
+      return originalGet!.call(this);
+    }
 
-  return computed({
-    get(key) {
-      return this.getAttribute(key);
-    },
-    set(key, value) {
+    function set(this: Model, value: any) {
```
Comment (dgeb) [inline addon/-private/fields/attr.ts:30]:
```suggestion
    function set(record: Model, value: any) {
```

File: addon/-private/fields/attr.ts [hunk 4]
```diff
@@ -1,24 +1,62 @@
...
+      'orbit:notifier',
+      (record: Model) => setDefaultValue(record),
+      target,
+      key
+    );
+
+    return trackedDesc;
+  }
+
+  if (arguments.length === 3) {
```
Comment (dgeb) [inline addon/-private/fields/attr.ts:55]: What is the scenario in which this condition is needed?

File: addon/-private/model.ts [hunk 5]
```diff
@@ -168,69 +180,58 @@ export default class Model extends EmberObject {
...
-          type: meta.options.type,
-          inverse: meta.options.inverse,
-          dependent: meta.options.dependent
-        };
+    return this.getPropertiesMeta('relationship');
+  }
+
+  static getPropertiesMeta(kind: string) {
+    const properties = Object.getOwnPropertyNames(this.prototype);
+    const metas = {};
```
Comment (dgeb) [inline addon/-private/model.ts:219]:
```suggestion
    const meta = {};
```

I think that `meta`, like `data`, should technically be considered plural (and I'm not sure the `s` adds clarity). Sorry to be pedantic about naming!

File: addon/-private/model.ts [hunk 6]
```diff
@@ -20,11 +20,23 @@ interface HasManyContract {
...
+}
+
+export default class Model {
+  static _notifiers: Dict<(instance: Model) => void> = {};
+
+  readonly identity!: RecordIdentity;
 
-  private _store?: Store;
-  private _relatedRecords: Dict<HasManyContract> = {};
+  #store?: Store;
```
Comment (dgeb) [inline addon/-private/model.ts:33]: Nice to finally use private fields!

File: addon/-private/fields/attr.ts (hunk 2 above)
Comment (tchak) [reply addon/-private/fields/attr.ts:23]: `this` is not argument name but special TS thing to provide `this` type

Comment (tchak) [reply addon/-private/fields/attr.ts:23]: In this case `get` takes 0 arguments but it is executed on the object of type `Model`

Comment (dgeb) [reply addon/-private/fields/attr.ts:23]: Ah, thanks for the clarification! This made no sense to me from a JS perspective, but I can now see what TS is doing.

File: addon/-private/fields/attr.ts (hunk 4 above)
Comment (tchak) [reply addon/-private/fields/attr.ts:55]: It is the cas of `@attr() name;` vs `@attr name;`

File: addon/-private/fields/attr.ts (hunk 1 above)
Comment (tchak) [reply addon/-private/fields/attr.ts]: yes

**Comment by tchak** [comment]
Updated with fixes

**Pull request merged**